File: ProblemSolving/groupAnagrams/group.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        
        # Logic 1: 100 pass 80% faster - Instead of using collections for each word lets use sorted of each word and use them as key for dictionarys
        anagrams = {}
        for word in strs:
            key = "".join(sorted(word))
            if key not in anagrams:
                anagrams[key] = []
            anagrams[key].append(word)
        result = []
        for key, value in anagrams.items():
            result.append(value)
        return result
            

File: ProblemSolving/groupAnagrams/group.go
import (
    "fmt"
)

// Logic 1: Using to create a char freq map for every string and use the map as key for another map to track similar anagrams ( implement something kind of like set intersection )
// Can we use map as key to another map?
/*
func convertStringToMap(s string) map[rune]int {
    charMap := make(map[rune]int)
    for _, ch := range s {
        charMap[ch] += 1
    }
    return charMap
}
*/

// Logic 2: Sort the characters of a string and use them as key to the main anagram map

func sortCharInString(s string) string {
    split := strings.Split(s, "")
    sort.Strings(split)
    return strings.Join(split, "")
}


func groupAnagrams(strs []string) [][]string {
    groups := make(map[string][]string)
    for _, s := range strs {
        //sMap := convertStringToMap(s)
        sortStr := sortCharInString(s)
        groups[sortStr] = append(groups[sortStr], s)
    }
    anagrams := [][]string{}
    for _, value := range groups {
        anagrams = append(anagrams, value)
    }
    return anagrams
}
